src/interp/interp-wasi.cc

```diff
@@ -652,85 +653,90 @@ std::unordered_map<Instance*, WasiInstance*> wasiInstances;
 namespace wabt {
 namespace interp {
 
-Result WasiBindImports(const Module::Ptr& module,
-                       RefVec& imports,
-                       Stream* stream,
-                       Stream* trace_stream) {
-  Store* store = module.store();
-  for (auto&& import : module->desc().imports) {
-    if (import.type.type->kind != ExternKind::Func) {
-      stream->Writef("wasi error: invalid import type: %s\n",
-                     import.type.name.c_str());
-      return Result::Error;
+Result WasiRegisterInstance(const Instance::Ptr& instance,
+                            uvwasi_s* uvwasi,
+                            Stream* err_stream,
+                            Stream* trace_stream) {
+  Store* store = instance.store();
+  auto module = store->UnsafeGet<Module>(instance->module());
+  auto&& module_desc = module->desc();
+  
+  Memory::Ptr memory;
+  for (auto&& export_ : module_desc.exports) {
+    if (export_.type.name == "memory") {
+      if (export_.type.type->kind != ExternalKind::Memory) {
+        err_stream->Writef("wasi error: memory export has incorrect type\n");
+        return Result::Error;
+      }
+      memory = store->UnsafeGet<Memory>(instance->memories()[export_.index]);
+      break;
     }
+  }
 
-    if (import.type.module != "wasi_snapshot_preview1" &&
-        import.type.module != "wasi_unstable") {
-      stream->Writef("wasi error: unknown module import: `%s`\n",
-                     import.type.module.c_str());
-      return Result::Error;
-    }
+  if (!memory) {
+    err_stream->Writef("wasi error: memory export not found\n");
+    return Result::Error;
+  }
+
+  WasiInstance* wasi = new WasiInstance(instance, uvwasi, std::move(memory).get(), trace_stream);
+  wasiInstances[instance.get()] = std::move(wasi);
```

Comment:
I don't think `std::move` is meaningful on raw pointer.

Maybe just avoid the local and do `wasiInstances[instance.get()] = new WasiInstance...`